File: quote/Projects/PluginModel/template/Customer.vb
Imports System.Globalization
Imports DB

Namespace Template

    ''' <summary>
    ''' Represents the customer
    ''' </summary>
    ''' <remarks></remarks>
    Public Class Customer

        Private _Name As String
        Private _ID As Integer

        Public ReadOnly Property Name As String
            Get
                Return _Name
            End Get
        End Property

        Public Sub SetName(name As String)
            _Name = name
        End Sub

        Public ReadOnly Property Id As String
            Get
                Return _ID
            End Get
        End Property

        Public Sub SetId(id As Integer)
            _ID = id
        End Sub

        Public Overrides Function GetHashCode() As Integer
            Return _ID.GetHashCode()
        End Function

        Public Overrides Function Equals(obj As Object) As Boolean
            Dim other As Customer = obj
            Return (_ID = other._ID)
        End Function

        Public Overrides Function ToString() As String
            If _ID = 0 Then
                Return _Name
            End If
            Return _ID & " " & _Name
        End Function
        
        Public Shared Function GetById(id As Integer) As Customer

            Dim customer As New Customer
            Dim adaptor As New QuoteDataBaseTableAdapters.CustomerTableAdapter
            Dim table As QuoteDataBase.CustomerDataTable
            table = adaptor.GetData()
            
            For Each row As QuoteDataBase.CustomerRow In table.Rows
                If row.CustomerID = id Then
                    customer.SetID(row.CustomerID)
                    customer.SetName(row.CustomerName.Trim())
                    Return customer
                End If
            Next

            Return Nothing
        End Function

        Public Shared Function GetByName(name As String) As Customer

            Dim customer As New Customer
            Dim adaptor As New QuoteDataBaseTableAdapters.CustomerTableAdapter
            Dim table As QuoteDataBase.CustomerDataTable
            table = adaptor.GetData()
            For Each row As QuoteDataBase.CustomerRow In table.Rows
                Dim cname As String = row.CustomerName.ToLowerInvariant().Trim()
                If cname = name.Trim() Then
                    customer.SetName(row.CustomerName.Trim())
                    customer.SetID(row.CustomerID)
                    Return customer
                End If
            Next

            Return Nothing
        End Function

        Public Shared Function CreateFromString(value As String) As Customer

            Dim index As Integer
            index = value.IndexOf(" ")

            If index = -1 Then
                Dim c As New Customer
                c.SetName(value)
                Return c
            End If

            Dim left As String = value.Substring(0, index)
            Dim right As String = value.Substring(index)

            Dim customer As New Template.Customer
            If left.Length > 0 Then
                Dim id As Integer
                Integer.TryParse(left, id)
                Dim name As String = right.Trim()

                customer.SetID(id)
                customer.SetName(name)
            Else
                customer.SetName(value.ToString())
            End If

            Return customer

        End Function

    End Class
End Namespace

File: trunk/quote/Projects/PluginModel/template/Customer.cs
namespace Model.Template
{
    using System;
    using System.Linq;

    /// <summary>
    /// Represents the customer
    /// </summary>
    /// <remarks></remarks>
    public class Customer
    {
        private string name;

        private int id;
        
        public string Name 
        {
            get { return this.name; }
        }

        public int Id 
        {
            get { return this.id; }
        }

        public static Customer GetById(int id)
        {
            Customer customer = new Customer();
            DB.QuoteDataBaseTableAdapters.CustomerTableAdapter adaptor = new DB.QuoteDataBaseTableAdapters.CustomerTableAdapter();
            DB.QuoteDataBase.CustomerDataTable table = null;
            table = adaptor.GetData();

            foreach (DB.QuoteDataBase.CustomerRow row in table.Rows) 
            {
                if (row.CustomerID == id) 
                {
                    customer.SetId(row.CustomerID);
                    customer.SetName(row.CustomerName.Trim());
                    return customer;
                }
            }

            return null;
        }

        public static Customer GetByName(string name)
        {
            Customer customer = new Customer();
            DB.QuoteDataBaseTableAdapters.CustomerTableAdapter adaptor = new DB.QuoteDataBaseTableAdapters.CustomerTableAdapter();
            DB.QuoteDataBase.CustomerDataTable table = null;
            table = adaptor.GetData();
            
            foreach (DB.QuoteDataBase.CustomerRow row in table.Rows) 
            {
                string cname = row.CustomerName.ToUpperInvariant().Trim();
                
                if (cname == name.ToUpperInvariant().Trim()) 
                {
                    customer.SetName(row.CustomerName.Trim());
                    customer.SetId(row.CustomerID);
                    return customer;
                }
            }

            return null;
        }

        public static Customer CreateFromString(string value)
        {
            int index = 0;
            index = value.IndexOf(" ", StringComparison.CurrentCulture);

            if (index == -1) 
            {
                Customer c = new Customer();
                c.SetName(value);
                return c;
            }

            string left = value.Substring(0, index);
            string right = value.Substring(index);

            Template.Customer customer = new Template.Customer();
            if (left.Length > 0) 
            {
                int id = 0;

                var parseOk = int.TryParse(left, out id);
                if (!parseOk) 
                {
                    id = 0;
                }

                string name = right.Trim();

                customer.SetId(id);
                customer.SetName(name);
            } 
            else 
            {
                customer.SetName(value.ToString());
            }

            return customer;
        }
        
        public void SetName(string name)
        {
            this.name = name;
        }

        public void SetId(int id)
        {
            this.id = id;
        }

        public override int GetHashCode()
        {
            return this.id.GetHashCode();
        }

        public override bool Equals(object obj)
        {
            Customer other = obj as Customer;
            return this.id == other.id;
        }

        public override string ToString()
        {
            if (this.id == 0) 
            {
                return this.name;
            }
            
            return this.id + " " + this.name;
        }
    }
}
